cpp/include/cudf/aggregation.hpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <vector>
+
+/**
+ * @file aggregation.hpp
+ * @brief Representation for specifying desired aggregations from
+ * aggregation-based APIs, e.g., groupby, reductions, rolling, etc.
+ *
+ * @note Not all aggregation APIs support all aggregation operations. See
+ * individual function documentation to see what aggregations are supported.
+ *
+ */
+
+namespace cudf {
+namespace experimental {
+/**
+ * @brief Base class for abstract representation of an aggregation.
+ */
+class aggregation;
+
+/// Factory to create a SUM aggregation
+std::unique_ptr<aggregation> make_sum_aggregation();
```

Comment:
Why do we need a factory for every trivial aggregation type? Seems to me the reason for this complexity is due to one outlier: quantiles. Why not treat quantiles specially rather than making everything more complex? Just use an enum class for all the simple aggregations, and complex aggregations can be separate. (?)

Reply:
> Seems to me the reason for this complexity is due to one outlier: quantiles.

For _now_, sure, but as we add more complex aggregations, factories with arguments will become the norm rather than the outlier. For example, Spark wants a "count" aggregation where you can parameterize whether or not null values are counted.

> Just use an enum class for all the simple aggregations, and complex aggregations can be separate.

All aggregations need to share a common base class in order to store all possible aggregations in a container like `std::vector<std::unique_ptr<aggregation>>`. The _derived_ classes are separate, but they all need a common representation.

Reply:
But do we need factories for every one, or would a constructor suffice?

Reply:
What would the constructor look like? Probably something like:
```
aggregation(aggregation::Kind kind);
```

but then you'd have to document and error check for all the values of `kind` that are acceptable (e.g., only `min/max/sum`,etc). Plus `aggregation` could no longer be an opaque type. 

Furthermore, it has to be a `unique_ptr`, so users would have to do something like:
```
make_unique<aggregation>(...);
```

It's not really any more code to just use a factory, and the factory is explicit and unambiguous.

Reply:
This is why I asked. I didn't know the things you describe. Thanks for explaining so patiently. :-)